[2/k] Provide repo load data contextually when loading user code

benpankow opened the pull request:
## Summary & Motivation

## How I Tested These Changes

benpankow commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23341?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#23341** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23341?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#23307** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23307?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

OwenKephart commented on python_modules/dagster/dagster_tests/core_tests/pending_repo_from_cached_asset_metadata.py:
```diff
@@ -0,0 +1,76 @@
+@asset(
+    metadata={
+        CACHED_ASSET_ID_KEY: "my_cached_asset_id",
+        CACHED_ASSET_METADATA_KEY: {"foo": "bar"},
```
so something that threw me for a bit more of a loop than it should have is the fact that this metadata is hardcoded on these assets, and so the stuff that's loaded from the repository load data is irrelevant.

I can see that the tests below are still valid, but I think this would make a bit more sense if you just wrote a function that returns assets based on the cached metadata, just to be a bit more "realistic"

benpankow replied:
Updated

OwenKephart reviewed (approved):
Note comment on test

schrockn reviewed (changes requested):
Per our live discussion, I have pretty major concerns about an approach that is this invasive that ends up with import-time fetches and a global context var object. I don't think this will be easily revertible.

Let's try to workaround some of the oddness with a lighter approach.

In the end in order to do this truly cleanly we will need a new code location API that allows the creation of a Definitions object in a callback

benpankow commented:
For now, moved the integration stack to use existing cacheable assets abstraction. Updated this PR to tuck the "fetch cached data" fn in the instance as discussed.

#23253 moves us in the direction of lazy loading definitions, and would let us attach this information to the load context.
